### PR TITLE
added support for component Settings UI

### PR DESCRIPTION
**Pull request by omkarkhatavkar**
### PR Objective 
This will add support for Settings UI component from Satellite. 

### Test Result 

```
(env) [root@okhatavk robottelo]# pytest tests/foreman/ui_airgun/test_settings.py -v
========================================================= test session starts ==========================================================
platform linux -- Python 3.4.8, pytest-4.0.2, py-1.7.0, pluggy-0.9.0 -- /home/okhatavk/Satellite/robottelo/env/bin/python3.4
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/okhatavk/Satellite/robottelo, inifile:
plugins: xdist-1.26.1, services-1.3.1, mock-1.10.0, forked-1.0.2, cov-2.6.1
collecting ... 2019-06-04 11:24:11 - conftest - DEBUG - Collected 24 test cases

collected 24 items                                                                                                                     

tests/foreman/ui_airgun/test_settings.py::test_positive_update_login_page_footer_text_with_long_string SKIPPED                   [  4%]
tests/foreman/ui_airgun/test_settings.py::test_positive_remove_login_page_footer_text SKIPPED                                    [  8%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_authorize_login_delegation_param PASSED                           [ 12%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_administrator_param PASSED                                        [ 16%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_authorize_login_delegation_api_param PASSED                       [ 20%]
tests/foreman/ui_airgun/test_settings.py::test_negative_update_entries_per_page_param PASSED                                     [ 25%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_entries_per_page_param PASSED                                     [ 29%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_email_reply_address_param PASSED                                  [ 33%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_fix_db_cache_param PASSED                                         [ 37%]
tests/foreman/ui_airgun/test_settings.py::test_negative_update_max_trend_param PASSED                                            [ 41%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_max_trend_param PASSED                                            [ 45%]
tests/foreman/ui_airgun/test_settings.py::test_negative_update_idle_timeout_param PASSED                                         [ 50%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_idle_timeout_param PASSED                                         [ 54%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_foreman_url_param PASSED                                          [ 58%]
tests/foreman/ui_airgun/test_settings.py::test_negative_update_foreman_url_param PASSED                                          [ 62%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_login_page_footer_text PASSED                                     [ 66%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_dynflow_enable_console_param PASSED                               [ 70%]
tests/foreman/ui_airgun/test_settings.py::test_negative_settings_access_to_non_admin SKIPPED                                     [ 75%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_email_delivery_method_smtp SKIPPED                                [ 79%]
tests/foreman/ui_airgun/test_settings.py::test_negative_update_email_delivery_method_smtp SKIPPED                                [ 83%]
tests/foreman/ui_airgun/test_settings.py::test_positive_update_email_delivery_method_sendmail SKIPPED                            [ 87%]
tests/foreman/ui_airgun/test_settings.py::test_negative_update_email_delivery_method_sendmail SKIPPED                            [ 91%]
tests/foreman/ui_airgun/test_settings.py::test_positive_email_yaml_config_precedence SKIPPED                                     [ 95%]
tests/foreman/ui_airgun/test_settings.py::test_negative_update_hostname_with_empty_fact SKIPPED                                  [100%]


====================================== 15 passed, 9 skipped in 2395.24 seconds ======================================

```
### Note 
Corresponding robottelo pr is added here : https://github.com/SatelliteQE/robottelo/pull/6991

### CODE REVIEW

File: airgun/entities/settings.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from airgun.entities.base import BaseEntity
```
Comment (oshtaier) [inline airgun/entities/settings.py]: Here is my vision of implementation
```
from airgun.entities.base import BaseEntity
from airgun.navigation import NavigateStep, navigator
from airgun.views.settings import SettingEditView, SettingsView


class SettingsEntity(BaseEntity):
    def search(self, value):
        view = self.navigate_to(self, 'All')
        return view.search(value)

    def read(self, property_name):
        """Read settings values"""
        view = self.navigate_to(self, 'All')
        view.search(property_name)
        return view.read()

    def update(self, property_name, value):
        """Update setting property with provided value"""
        view = self.navigate_to(self, 'All')
        view.search(property_name)
        view.table.row()['Value'].widget.fill(value)

@navigator.register(SettingsEntity, 'All')
class ShowAllSettings(NavigateStep):
    """Navigate to All Settings page"""
    VIEW = SettingsView

    def step(self, *args, **kwargs):
        self.view.menu.select('Administer', 'Settings')
```

- Read doesn't need widgets_name as you have only one widget here
- You don't need extra navigation step for Edit as you actually do not navigate anywhere, but stay on 'All'
- You can't just add such `menu...` post actions as they looks completely out of style and redundant. You do `search` for every action, so you get actual data all the time, so not sure why you need it at all. As last resort we can look for refresh, but still don't know why you need it
- We do not work with entities fill like this. Your widget `fill` method should know what you need: that you need click on widget, then fill, then click `submit`, etc. So everything implemented directly in view

File: airgun/views/settings.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+from wait_for import wait_for
+from widgetastic.widget import Text
+
+from airgun.views.common import BaseLoggedInView, SearchableViewMixin
+from airgun.widgets import PopOverWidget, SatTable
+
+
+class SettingsView(BaseLoggedInView, SearchableViewMixin):
+    title = Text("//h1[text()='Settings']")
+    table = SatTable(
```
Comment (oshtaier) [inline airgun/views/settings.py]: `SatTable` is deprecated, please use `Table` instead

File: airgun/views/settings.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        './/table',
+        column_widgets={
+            'Name': Text('.//a'),
```
Comment (oshtaier) [inline airgun/views/settings.py]: That is not true, Name is not a link

File: airgun/views/settings.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            'Value': Text('.//span[contains(@class, "editable-click")]')
```
Comment (oshtaier) [inline airgun/views/settings.py]: There is only one `span` element in that `td`, so that line is redundant

File: airgun/views/settings.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        },
+    )
+
+    @property
+    def is_displayed(self):
+        return self.browser.wait_for_element(
+            self.title, exception=False) is not None
+
+
+class SettingEditView(SettingsView):
```
Comment (oshtaier) [inline airgun/views/settings.py]: There is no Edit View for that entity

File: airgun/views/settings.py (hunk 4 above)
Comment (oshtaier) [reply airgun/views/settings.py]: or it is not, in case you like to use your custom widget

File: airgun/views/settings.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            'Value': PopOverWidget(
+                './/span[contains(@class, "editable-click")]')
+        },
+    )
+
+    def wait_for_update(self):
+        """Wait for value to update"""
+        wait_for(
+            lambda: not self.table.row()['Value'].widget.header.is_displayed,
+            timeout=10,
```
Comment (oshtaier) [inline airgun/views/settings.py]: probably that timeout is too small

File: airgun/widgets.py [hunk 7]
```diff
@@ -2090,3 +2091,42 @@ class Link(Text):
...
+    Locator example::
+
+        //div[contains(@class, 'editable-open')]
+    """
+
+    header = Text("//h3[@class='popover-title']")
+    input_box = TextInput(locator="//div[@class='editable-input']/input")
+    drop_down = Select("//div[@class='editable-input']/select")
+    submit = Text("//button[@type='submit']")
+
```
Comment (oshtaier) [inline airgun/widgets.py]: I would say that you need to implement `read` too

File: airgun/widgets.py [hunk 8]
```diff
@@ -2090,3 +2091,42 @@ class Link(Text):
...
+                <input type="text" ...>
+                <div class="editable-buttons">
+                    <button type="submit" class="btn btn-primary btn-sm editable-submit">
+
+    Locator example::
+
+        //div[contains(@class, 'editable-open')]
+    """
+
+    header = Text("//h3[@class='popover-title']")
```
Comment (oshtaier) [inline airgun/widgets.py]: all locators have to be relative:
```
    header = Text(".//h3[@class='popover-title']")
    input_box = TextInput(locator=".//div[@class='editable-input']/input")
    drop_down = Select(".//div[@class='editable-input']/select")
    submit = Text(".//button[@type='submit']")
```

File: airgun/widgets.py [hunk 9]
```diff
@@ -2090,3 +2091,42 @@ class Link(Text):
...
+    input_box = TextInput(locator="//div[@class='editable-input']/input")
+    drop_down = Select("//div[@class='editable-input']/select")
+    submit = Text("//button[@type='submit']")
+
+    def fill(self, item):
+        """Selects value from drop_down if exist otherwise write into input_box"""
+        if self.header.is_displayed:
+            if self.drop_down.is_displayed:
+                self.drop_down.select_by_visible_text(item)
```
Comment (oshtaier) [inline airgun/widgets.py]: always use fill for all widgets `self.drop_down.fill()`

File: airgun/widgets.py [hunk 10]
```diff
@@ -2090,3 +2091,42 @@ class Link(Text):
...
+                <div class="editable-buttons">
+                    <button type="submit" class="btn btn-primary btn-sm editable-submit">
+
+    Locator example::
+
+        //div[contains(@class, 'editable-open')]
+    """
+
+    header = Text("//h3[@class='popover-title']")
+    input_box = TextInput(locator="//div[@class='editable-input']/input")
```
Comment (oshtaier) [inline airgun/widgets.py]: I feel that `TextInput(locator="//div[@class='editable-input']/input")` will not match: `<div class="editable-input"><textarea class="form-control input-large" placeholder="fqdn,hostgroup,os,domain" rows="7"></textarea></div>`
We need to be able fill both inputs and textareas here

File: airgun/entities/settings.py (hunk 1 above)
Comment (omkarkhatavkar) [reply airgun/entities/settings.py]: Excellent @oshtaier!

File: airgun/widgets.py (hunk 10 above)
Comment (omkarkhatavkar) [reply airgun/widgets.py]: nice catch, I  saw that but somehow missed it.

**Comment by omkarkhatavkar** [comment]
> Done with first round of review here. Pending comments

@oshtaier Please go ahead with the second round of review :)

File: airgun/widgets.py [hunk 11]
```diff
@@ -2090,3 +2091,48 @@ class Link(Text):
     def fill(self, value):
         if value:
             self.browser.click(self)
+
+
+class PopOverWidget(GenericLocatorWidget):
```
Comment (oshtaier) [inline airgun/widgets.py]: 1) Don't think that we need to put in basic widget module as it specific to settings module only
2) Can we think about better name?

File: airgun/widgets.py [hunk 12]
```diff
@@ -2090,3 +2091,48 @@ class Link(Text):
...
+            <form class="form-inline editableform" style="">
+                <input type="text" ...>
+                <div class="editable-buttons">
+                    <button type="submit" class="btn btn-primary btn-sm editable-submit">
+
+    Locator example::
+
+        //div[contains(@class, 'editable-open')]
+    """
+    column_text = Text("//span[contains(@class, 'editable-click')]")
```
Comment (oshtaier) [inline airgun/widgets.py]: maybe some other name too?

File: airgun/widgets.py [hunk 13]
```diff
@@ -2090,3 +2091,48 @@ class Link(Text):
...
+    header = Text("//*[contains(@class, 'popover-title')]")
```
Comment (oshtaier) [inline airgun/widgets.py]: Try to avoid `//*` if you can specify exact locator

File: airgun/widgets.py [hunk 14]
```diff
@@ -2090,3 +2091,48 @@ class Link(Text):
...
+    input_box = TextInput(locator="//*[contains(@class, 'form-control input-sm') or "
+                                  "contains(@class, 'form-control input-large')]")
+    drop_down = Select("//*[contains(@class, 'editable')]/select")
```
Comment (oshtaier) [inline airgun/widgets.py]: Try to avoid `//*` if you can specify exact locator

File: airgun/widgets.py [hunk 15]
```diff
@@ -2090,3 +2091,48 @@ class Link(Text):
...
+    submit = Text("//*[@type='submit']")
```
Comment (oshtaier) [inline airgun/widgets.py]: Try to avoid `//*` if you can specify exact locator

File: airgun/views/settings.py [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+from airgun.views.common import BaseLoggedInView, SearchableViewMixin
+from airgun.widgets import PopOverWidget, Table
+
+
+class SettingsView(BaseLoggedInView, SearchableViewMixin):
+    title = Text("//h1[text()='Settings']")
+    table = Table(
+        './/table',
+        column_widgets={
+            'Name': Text('.//span[contains(@rel,"twipsy")]'),
```
Comment (oshtaier) [inline airgun/views/settings.py]: It is not necessary to specify locator for every column. Do it only if you like to work with widgets inside

File: airgun/views/settings.py (hunk 6 above)
Comment (oshtaier) [reply airgun/views/settings.py]: so maybe 30?

File: airgun/widgets.py (hunk 11 above)
Comment (omkarkhatavkar) [reply airgun/widgets.py]: 1. One point 1, tomorrow it could have similar widget then? (All the actions inside that widget are specifically are common like widget nothing related settings module ) 
2. What would you think, give me the suggested name?

File: airgun/widgets.py (hunk 12 above)
Comment (omkarkhatavkar) [reply airgun/widgets.py]: suggestions?

File: airgun/widgets.py (hunk 13 above)
Comment (omkarkhatavkar) [reply airgun/widgets.py]: I changed because of your comment mentioned here https://github.com/SatelliteQE/airgun/pull/334#discussion_r290263646. Do you mean I wanted to use tag instead of '*' and use contains more likely?

File: airgun/views/settings.py (hunk 16 above)
Comment (omkarkhatavkar) [reply airgun/views/settings.py]: Hmm, that's a valid point. I will keep it in mind as this will automatically be known by the table

File: airgun/widgets.py (hunk 13 above)
Comment (oshtaier) [reply airgun/widgets.py]: I mean relative paths `.//`, not the `//*` =)

**Comment by oshtaier** [comment]
Hey @omkarkhatavkar  Do you need some explanation on what `.//` mean? That `.` in `.//`?

**Comment by oshtaier** [comment]
Just basically, it is `TOP_VIEW_ROOT_LOCATOR/NEXT_VIEW_ROOT_LOCATOR/WIDGET_ROOT_LOCATOR/.../ROOT`

**Comment by omkarkhatavkar** [comment]
> Hey @omkarkhatavkar Do you need some explanation on what `.//` mean? That `.` in `.//`?

I knew is `'//'` mainly used while searching anywhere in whole HTML tree and used for relative XPath. I did not know about `'.'`.

**Comment by oshtaier** [comment]
@omkarkhatavkar  it is very important idea, that is need to understand.
It is not about xpath rules, but thing that implemented from widgetastic standpoint.

If you say `'Value': PopOverWidget('//span[contains(@class, "editable-click")]')` it will look for `//span[contains(@class, "editable-click")]`

If you say `'Value': PopOverWidget('.//span[contains(@class, "editable-click")]')` it will look for ROOT for `//span[contains(@class, "editable-click")]` and add it to locator. In that particular example, ROOT for table column is table itself, so we get to `.//table//span[contains(@class, "editable-click")]`. And if that table is in tab, so its tab locator will be added to table `.//tab_locator//table//span[contains(@class, "editable-click")]`.
So basically `.//span[contains(@class, "editable-click")]` is refer to `//ROOT//tab_locator//table//span[contains(@class, "editable-click")]`. In that way, it is really easy to find unique elements. You don't need to write '//span[very long and complex locator]', but just './/span' and it will look for exact span relatively to your ROOT element

**Comment by omkarkhatavkar** [comment]
> @omkarkhatavkar it is very important idea, that is need to understand.
> It is not about xpath rules, but thing that implemented from widgetastic standpoint.
> 
> If you say `'Value': PopOverWidget('//span[contains(@class, "editable-click")]')` it will look for `//span[contains(@class, "editable-click")]`
> 
> But if you say `'Value': PopOverWidget('.//span[contains(@class, "editable-click")]')` it will look for ROOT for `//span[contains(@class, "editable-click")]` and add it to locator. In that particular example, ROOT for table column is table itself, so we get to `.//table//span[contains(@class, "editable-click")]`. And if that table is in tab, so its tab locator will be added to table `.//tab_locator//table//span[contains(@class, "editable-click")]`. And so on
> So basically `.//span[contains(@class, "editable-click")]` is refer to `//ROOT//tab_locator//table//span[contains(@class, "editable-click")]`. In that way, it is really easy to find unique elements. You don't need to write '//span[very long and complex locator]', but just './/span' and it will look for exact span relatively to your ROOT element

Thanks for that explanation, frankly I was not aware of it. I was reading https://www.bennadel.com/blog/2142-using-and-expressions-in-xpath-xml-search-directives-in-coldfusion.htm understand that. But your explanation was spot on.  Thanks for that.

**Comment by omkarkhatavkar** [comment]
@oshtaier I'm getting an error while using `".//span[contains(@class, 'editable-click')]"`  and `'.//span[contains(@class, "editable-click")]'` where as earlier `'//span[contains(@class, "editable-click")]'` was working. 

### Error 
#### 1. Error with `".//span[contains(@class, 'editable-click')]"` locator 

```
self = <BrowserParentWrapper for PopOverWidget('.//span[contains(@class, "editable-click")]')>
locator = Locator(by='xpath', locator=".//span[contains(@class, 'editable-click')]")
args = (), kwargs = {}, vcheck = None, elements = []

selenium.common.exceptions.NoSuchElementException: Message: Could not find an element Locator(by='xpath', locator=".//span[contains(@class, 'editable-click')]")

```

#### 2. Error with './/span[contains(@class, "editable-click")]' locator 
```
self = <BrowserParentWrapper for PopOverWidget('.//span[contains(@class, "editable-click")]')>
locator = Locator(by='xpath', locator='.//span[contains(@class, "editable-click")]')
args = (), kwargs = {}, vcheck = None, elements = []

E           selenium.common.exceptions.NoSuchElementException: Message: Could not find an element Locator(by='xpath', locator='.//span[contains(@class, "editable-click")]')

```

### Note 
Even if I use`.//` in front other locators it seems to be not working in this case. @oshtaier  Please help me to write good locators in this case.

**Comment by oshtaier** [comment]
Did you do?:
```
        column_widgets={
            'Value': PopOverWidget(
                './/span[contains(@class, "editable-click")]')
        },
```
and then:
```
    column_value = Text(".//span[contains(@class, 'editable-click')]")
```
Of course, that will never work as in result you will have:
`.//td//span[contains(@class, "editable-click")]//span[contains(@class, 'editable-click')]`
Could you please try:
```
        column_widgets={
            'Value': PopOverWidget('.')
        }

    column_value = Text(".//span[contains(@class, 'editable-click')]")
```
If that works maybe you use:
`class PopOverWidget(Widget):`
instead
`class PopOverWidget(GenericLocatorWidget):`

**Comment by omkarkhatavkar** [comment]
@oshtaier worked all relative xpaths, come to know my obvious mistake. Although inheriting `Widget` did not work for me. That should not be a blocker issue, right?

**Comment by oshtaier** [comment]
No, it is blocking, as we do not need provide locators where it is not needed. Do you use `'Value': PopOverWidget()`? without locator?

**Comment by oshtaier** [comment]
or what is error?

**Comment by omkarkhatavkar** [comment]
**with class widget and 'Value': PopOverWidget()**

```
        try:
>           locator = self.__locator__()
E           AttributeError: 'PopOverWidget' object has no attribute '__locator__'
```
**with class widget and 'Value': PopOverWidget(".")**

```
>       return method(self, *new_args, **new_kwargs)
E       TypeError: __init__() got multiple values for argument 'logger'

../../../env/lib/python3.4/site-packages/widgetastic/widget/base.py:67: TypeError
```

File: airgun/widgets.py [hunk 17]
```diff
@@ -2090,3 +2091,50 @@ class Link(Text):
...
+        self.column_value.click()
+        if self.header.is_displayed:
+            if self.drop_down.is_displayed:
+                self.drop_down.fill(item)
+            else:
+                self.input_box.fill(item)
+        self.submit.click()
+
+    def read(self):
+        """read colum updated value"""
```
Comment (oshtaier) [inline airgun/widgets.py]: Typo `"""Read column updated value"""`

Comment (omkarkhatavkar) [reply airgun/widgets.py]: done !
